File: forecast/app/services/metrics.py
"""Allowlist of metrics the forecast service knows how to aggregate.

Mirrors api/src/events/metrics.ts — keep them in lockstep when adding new
metrics. Lives in a constant rather than a config file so the BFF and
the forecast service ship the same set of names without a runtime
handshake.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal


AggregationKind = Literal["sum_revenue", "count_events", "unique_users"]


@dataclass(frozen=True, slots=True)
class MetricDef:
    label: str
    event_name: str
    agg: AggregationKind
    unit: Literal["cents", "count", "users"]


METRICS: dict[str, MetricDef] = {
    "mrr": MetricDef("Monthly recurring revenue", "subscription_payment", "sum_revenue", "cents"),
    "dau": MetricDef("Daily active users", "session_start", "unique_users", "users"),
    "signups": MetricDef("Signups", "signup", "count_events", "count"),
    "churn": MetricDef("Churn events", "subscription_cancelled", "count_events", "count"),
    "payments": MetricDef("Payments", "subscription_payment", "count_events", "count"),
}


def is_known(name: str) -> bool:
    return name in METRICS


def resolve(name: str) -> MetricDef:
    if name not in METRICS:
        raise KeyError(f"unknown metric: {name}")
    return METRICS[name]

File: api/src/events/metrics.ts
/**
 * Allowlist of metric names the BFF will aggregate against ClickHouse.
 *
 * Every metric maps to an event_name we expect in the events table plus
 * the aggregation it requires. New metrics must be added here — there is
 * no dynamic SQL composition path. The point is to make a hostile user
 * controlling `?metric=` parameter unable to influence the SQL plan.
 */

export type AggregationKind = 'sum_revenue' | 'count_events' | 'unique_users';

export interface MetricDefinition {
  /** Display name shown in the model/widget metadata. */
  label: string;
  /** Event name we aggregate in ClickHouse. */
  eventName: string;
  /** Which aggregate to read off the kpi_hourly materialized view. */
  agg: AggregationKind;
  /** Currency for monetary metrics (informational; values stay in cents). */
  currency?: string;
}

export const METRICS: Readonly<Record<string, MetricDefinition>> = Object.freeze({
  mrr: {
    label: 'Monthly recurring revenue',
    eventName: 'subscription_payment',
    agg: 'sum_revenue',
    currency: 'USD',
  },
  dau: {
    label: 'Daily active users',
    eventName: 'session_start',
    agg: 'unique_users',
  },
  signups: {
    label: 'Signups',
    eventName: 'signup',
    agg: 'count_events',
  },
  churn: {
    label: 'Churn events',
    eventName: 'subscription_cancelled',
    agg: 'count_events',
  },
  payments: {
    label: 'Payments',
    eventName: 'subscription_payment',
    agg: 'count_events',
  },
});

export function isKnownMetric(name: string): name is keyof typeof METRICS {
  return Object.prototype.hasOwnProperty.call(METRICS, name);
}
